Source/WTF/wtf/Language.cpp

```diff
@@ -166,55 +167,66 @@ void overrideUserPreferredLanguages(const Vector<String>& override)
 
 static String canonicalLanguageIdentifier(const String& languageCode)
 {
-    String lowercaseLanguageCode = languageCode.convertToASCIILowercase();
-    
-    if (lowercaseLanguageCode.length() >= 3 && lowercaseLanguageCode[2] == '_')
-        lowercaseLanguageCode = makeStringByReplacing(lowercaseLanguageCode, 2, 1, "-"_s);
+    return makeStringByReplacingAll(languageCode.convertToASCIILowercase(), '_', '-');
+}
+
+LocaleComponents parseLocale(const String& localeIdentifier)
+{
+    LocaleComponents result;
+
+    auto components = canonicalLanguageIdentifier(localeIdentifier).split('-');
+    if (components.size() >= 1)
+        result.languageCode = components[0];
+
+    if (components.size() >= 2 && components[1].length() == 4)
+        result.scriptCode = makeString(components[1].left(1).convertToASCIIUppercase(), components[1].substring(1));
 
-    return lowercaseLanguageCode;
+    if (components.size() >= 3 && !result.scriptCode)
+        result.countryCode = components[1].convertToASCIIUppercase();
+    else if (components.size() >= 3)
+        result.countryCode = components[2].convertToASCIIUppercase();
+
+    return result;
```

Comment:
Is this locale string documented anywhere? It might be nice to link to the documentation to explain what exactly is happening here.

Reply:
I'll add a comment and link to [BCP47](https://www.ietf.org/rfc/bcp/bcp47.txt).